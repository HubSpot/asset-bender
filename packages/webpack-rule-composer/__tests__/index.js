/* @flow */

import type { RuleFn } from '../src';
import getRuleComposer from '../src';

describe('getRuleComposer', () => {
  const webpackOptions = {};
  const argv = {};
  const other = {};
  const args = [webpackOptions, argv, other];

  it('returns a function', () => {
    const ruleResolver = () => [];
    expect(typeof getRuleComposer(ruleResolver, ...args)).toBe('function');
  });

  describe('ruleComposer', () => {
    it('resolves string rule names to literal rules', () => {
      const ruleComposer = getRuleComposer(
        () => [{ test: /simple-string-result/ }],
        ...args
      );
      expect(ruleComposer('string')).toMatchSnapshot();
    });

    it('recursively resolves rule names to literal rules', () => {
      const ruleComposer = getRuleComposer(ruleName => {
        if (ruleName === 'recursive-lookup') {
          return ['look', 'harder', { test: /look-no-further/ }];
        } else if (ruleName === 'look') {
          return [{ test: /recursive-look-result/ }];
        } else {
          return [{ test: /recursive-harder-result/ }];
        }
      }, ...args);
      expect(ruleComposer('recursive-lookup')).toMatchSnapshot();
    });
  });

  it('supports functions that return literal rules', () => {
    const ruleComposer = getRuleComposer(() => [], ...args);
    expect(
      ruleComposer(() => [{ test: /simple-function-result/ }])
    ).toMatchSnapshot();
  });

  it('supports functions recursively', () => {
    const ruleComposer = getRuleComposer(() => [], ...args);
    expect(
      ruleComposer(() => () => [{ test: /recursive-function-result/ }])
    ).toMatchSnapshot();
  });

  it('recursively resolves mixed strings and functions', () => {
    const ruleComposer = getRuleComposer(ruleName => {
      if (ruleName === 'initial') {
        return ['foo', (() => 'foo': RuleFn), { test: /look-no-further/ }];
      } else if (ruleName === 'foo') {
        return [{ test: /i-show-up-twice/ }];
      } else {
        return [];
      }
    }, ...args);
    expect(ruleComposer('initial')).toMatchSnapshot();
  });
});
