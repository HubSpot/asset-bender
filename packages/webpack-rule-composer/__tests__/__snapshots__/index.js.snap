// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRuleComposer recursively resolves mixed strings and functions 1`] = `
Array [
  Object {
    "test": /i-show-up-twice/,
  },
  Object {
    "test": /i-show-up-twice/,
  },
  Object {
    "test": /look-no-further/,
  },
]
`;

exports[`getRuleComposer ruleComposer recursively resolves rule names to literal rules 1`] = `
Array [
  Object {
    "test": /recursive-look-result/,
  },
  Object {
    "test": /recursive-harder-result/,
  },
  Object {
    "test": /look-no-further/,
  },
]
`;

exports[`getRuleComposer ruleComposer resolves string rule names to literal rules 1`] = `
Array [
  Object {
    "test": /simple-string-result/,
  },
]
`;

exports[`getRuleComposer supports functions recursively 1`] = `
Array [
  Object {
    "test": /recursive-function-result/,
  },
]
`;

exports[`getRuleComposer supports functions that return literal rules 1`] = `
Array [
  Object {
    "test": /simple-function-result/,
  },
]
`;
